File: tests_wraps/src/main/java/com/github/klee0kai/test/car/di/lists/cached/CarMultiCachedComponent.java
package com.github.klee0kai.test.car.di.lists.cached;

import com.github.klee0kai.stone.annotations.component.Component;
import com.github.klee0kai.stone.wrappers.Ref;
import com.github.klee0kai.test.car.model.Bumper;
import com.github.klee0kai.test.car.model.Car;
import com.github.klee0kai.test.car.model.Wheel;
import com.github.klee0kai.test.car.model.Window;

import javax.inject.Provider;
import java.lang.ref.WeakReference;
import java.util.List;

@Component
public interface CarMultiCachedComponent {

    CarMultiCachedModule cachedModule();

    Ref<Bumper> singleBumper();

    List<Provider<WeakReference<Wheel>>> wheels();

    Wheel wheel();

    List<List<Window>> windows();

    List<Provider<List<Window>>> windowsProviding();

    List<Car> cars();

}

File: tests_wraps_kotlin/src/main/kotlin/com/github/klee0kai/test/car/di/lists/cached/CarMultiCachedComponent.kt
package com.github.klee0kai.test.car.di.lists.cached

import com.github.klee0kai.stone.annotations.component.Component
import com.github.klee0kai.stone.annotations.qualifier.IgnoreQualifier
import com.github.klee0kai.stone.weakref.Ref
import com.github.klee0kai.test.car.model.Bumper
import com.github.klee0kai.test.car.model.Car
import com.github.klee0kai.test.car.model.Wheel
import com.github.klee0kai.test.car.model.Window
import java.lang.ref.WeakReference
import javax.inject.Provider

@Component
interface CarMultiCachedComponent {

    fun cachedModule(): CarMultiCachedModule?

    fun singleBumper(): Ref<Bumper?>?

    @IgnoreQualifier
    fun wheels(): List<Provider<WeakReference<Wheel?>?>?>?

    fun wheel(): Wheel?

    @IgnoreQualifier
    fun windows(): List<List<Window?>?>?

    @IgnoreQualifier
    fun windowsProviding(): List<Provider<List<Window?>?>?>?

    fun cars(): List<Car?>?
}
